File: web/src/lib/enrichment.ts
// Presentation helpers for a job's AI enrichment. Pure functions that turn the
// controlled-vocabulary codes (validated server-side) into display labels.
// Unknown codes fall back to a humanized form so a future vocabulary addition
// never renders blank — the SPA never re-validates, it only formats.

import type { Enrichment, Job } from './types';

/** One value within a facet row: its display text and, when the facet maps to a
 *  job-search filter, the /jobs URL that applies it. */
export interface FacetValue {
  text: string;
  href?: string;
}

/** A labelled facet row. Most facets carry a single value; the array-valued ones
 *  (region, country, industry) carry one entry per code, each independently
 *  clickable. */
export interface Facet {
  label: string;
  values: FacetValue[];
}

const SENIORITY: Record<string, string> = {
  intern: 'Intern',
  junior: 'Junior',
  middle: 'Middle',
  senior: 'Senior',
  lead: 'Lead',
  principal: 'Principal',
  c_level: 'C-level',
};

const EMPLOYMENT: Record<string, string> = {
  full_time: 'Full-time',
  part_time: 'Part-time',
  contract: 'Contract',
  internship: 'Internship',
};

const WORK_MODE: Record<string, string> = {
  remote: 'Remote',
  hybrid: 'Hybrid',
  onsite: 'On-site',
};

// Region codes (the top-level `regions` facet) → readable labels. Unmapped codes humanize.
const REGION: Record<string, string> = {
  global: 'Global',
  eu: 'Europe',
  emea: 'EMEA',
  eea: 'EEA',
  uk: 'UK',
  americas: 'Americas',
  north_america: 'North America',
  latam: 'LATAM',
  apac: 'APAC',
  mena: 'MENA',
  africa: 'Africa',
  us: 'USA',
  ru: 'Russia',
  cis: 'CIS',
  central_asia: 'Central Asia',
};

const RELOCATION: Record<string, string> = {
  not_supported: 'Not supported',
  supported: 'Supported',
  required: 'Required',
};

const ENGLISH_LEVEL: Record<string, string> = {
  a1: 'A1',
  a2: 'A2',
  b1: 'B1',
  b2: 'B2',
  c1: 'C1',
  c2: 'C2',
  native: 'Native',
};

// Only the acronym/compound domains need a label; the rest (Crypto, Healthcare,
// Media…) humanize cleanly, so they fall through to `humanize` like CATEGORY.
const DOMAINS: Record<string, string> = {
  fintech: 'FinTech',
  ecommerce: 'E-commerce',
  saas: 'SaaS',
  gamedev: 'GameDev',
  edtech: 'EdTech',
  adtech: 'AdTech',
  govtech: 'GovTech',
};

const CATEGORY: Record<string, string> = {
  ml_ai: 'ML / AI',
  qa: 'QA',
  devops: 'DevOps',
  data_engineering: 'Data engineering',
  data_science: 'Data science',
  project_management: 'Project management',
};

const COMPANY_TYPE: Record<string, string> = {
  product: 'Product',
  startup: 'Startup',
  outsource: 'Outsource',
  outstaff: 'Outstaff',
  agency: 'Agency',
  inhouse: 'In-house',
  government: 'Government',
};

const CURRENCY_SYMBOL: Record<string, string> = { USD: '$', EUR: '€', GBP: '£' };

const PERIOD_SUFFIX: Record<string, string> = {
  month: ' / mo',
  day: ' / day',
  hour: ' / hr',
  // `year` is the implicit default and reads cleaner without a suffix.
};

/** Title-case an unknown snake_case code (e.g. "data_engineering" → "Data engineering"). */
function humanize(value: string): string {
  const spaced = value.replace(/_/g, ' ');
  return spaced.charAt(0).toUpperCase() + spaced.slice(1);
}

/** Look a code up in its label map, humanizing anything outside the map. */
function label(map: Record<string, string>, value: string): string {
  return map[value] ?? humanize(value);
}

/** The /jobs URL that filters by a single facet value. Param names match the
 *  search API (see facets.ts / filters.svelte.ts). */
export function filterHref(param: string, value: string): string {
  return `/jobs?${param}=${encodeURIComponent(value)}`;
}

/** Group thousands with thin spaces, matching the salary line in the design. */
function groupThousands(n: number): string {
  return n.toLocaleString('en-US').replace(/,/g, ' ');
}

/**
 * Render the compensation line, or null when no salary is stated. Handles the
 * full range, a min-only floor ("from …"), and a max-only ceiling ("up to …").
 * The currency symbol and period suffix trail the amount, as in the design.
 */
export function formatSalary(e: Enrichment): string | null {
  const { salary_min, salary_max } = e;
  if (salary_min == null && salary_max == null) return null;

  const symbol = e.salary_currency
    ? (CURRENCY_SYMBOL[e.salary_currency] ?? e.salary_currency)
    : '';
  const period = e.salary_period ? (PERIOD_SUFFIX[e.salary_period] ?? '') : '';
  const tail = `${symbol}${period}`;

  let amount: string;
  if (salary_min != null && salary_max != null) {
    amount = `${groupThousands(salary_min)} – ${groupThousands(salary_max)}`;
  } else if (salary_min != null) {
    amount = `from ${groupThousands(salary_min)}`;
  } else {
    amount = `up to ${groupThousands(salary_max as number)}`;
  }

  return tail ? `${amount} ${tail}` : amount;
}

/**
 * The work-arrangement label for compact contexts (list cards): the resolved
 * top-level `work_mode` (LLM value, else the one parsed from the location), or
 * null when neither stated it.
 */
export function workArrangement(job: Pick<Job, 'work_mode'>): string | null {
  return job.work_mode ? label(WORK_MODE, job.work_mode) : null;
}

/**
 * The job's geographic area as a concise label from the top-level `regions` —
 * e.g. `Global`, `Europe`, `USA`. Meaningful for any work mode (a remote role's
 * reach or an onsite role's area). Null when regions is unknown (empty is
 * unknown, not global).
 */
export function regionLabel(job: Pick<Job, 'regions'>): string | null {
  if (!job.regions?.length) return null;
  return job.regions.map((r) => label(REGION, r)).join(', ');
}

/**
 * The short tag row shown on a list card's header: work arrangement, region,
 * employment type, and grade — only those that are stated, in that order.
 * Compact by design (the full facet set lives on the detail page).
 */
export function cardTags(job: Job): string[] {
  const e = job.enrichment;
  const tags: string[] = [];

  const arrangement = workArrangement(job);
  if (arrangement) tags.push(arrangement);
  const region = regionLabel(job);
  if (region) tags.push(region);
  if (e?.employment_type) tags.push(label(EMPLOYMENT, e.employment_type));
  if (e?.seniority) tags.push(label(SENIORITY, e.seniority));

  return tags;
}

/**
 * Ordered facets for the summary meta-row. Only facets with a stated value are
 * included, so an empty enrichment yields an empty list and the row hides
 * entirely. Order moves from work arrangement → role → eligibility → company,
 * mirroring the reference layout.
 */
export function summaryFacets(job: Job): Facet[] {
  const e = job.enrichment ?? {};
  const facets: Facet[] = [];

  // A scalar facet that maps to a search filter: one clickable value.
  const link = (
    name: string,
    param: string,
    code: string | null | undefined,
    text: string | null | undefined,
  ) => {
    if (code && text) facets.push({ label: name, values: [{ text, href: filterHref(param, code) }] });
  };
  // An array facet (region/country/industry): one clickable value per code.
  const links = (
    name: string,
    param: string,
    codes: string[] | undefined,
    toText: (code: string) => string,
  ) => {
    if (codes?.length) {
      facets.push({ label: name, values: codes.map((c) => ({ text: toText(c), href: filterHref(param, c) })) });
    }
  };
  // A facet with no matching filter: plain, non-clickable text.
  const plain = (name: string, text: string | null | undefined) => {
    if (text) facets.push({ label: name, values: [{ text }] });
  };

  link('Work format', 'work_mode', job.work_mode, job.work_mode && label(WORK_MODE, job.work_mode));
  plain('Location', job.location);
  links('Region', 'regions', job.regions, (r) => label(REGION, r));
  link('Work type', 'employment_type', e.employment_type, e.employment_type && label(EMPLOYMENT, e.employment_type));
  link('Grade', 'seniority', e.seniority, e.seniority && label(SENIORITY, e.seniority));
  plain('Experience', e.experience_years_min != null ? `${e.experience_years_min}+ yrs` : null);
  link(
    'English',
    'english_level',
    e.english_level && e.english_level !== 'none' ? e.english_level : null,
    e.english_level && e.english_level !== 'none' ? label(ENGLISH_LEVEL, e.english_level) : null,
  );
  link('Category', 'category', e.category, e.category && label(CATEGORY, e.category));
  links('Country', 'countries', job.countries, (c) => c.toUpperCase());
  link('Relocation', 'relocation', e.relocation, e.relocation && label(RELOCATION, e.relocation));
  if (e.visa_sponsorship === true) {
    facets.push({ label: 'Visa', values: [{ text: 'Sponsored', href: filterHref('visa_sponsorship', 'true') }] });
  }
  link('Company', 'company_type', e.company_type, e.company_type && label(COMPANY_TYPE, e.company_type));
  plain('Size', e.company_size);
  links('Domains', 'domains', e.domains, (d) => label(DOMAINS, d));

  return facets;
}

File: internal/enrich/enrichment.go
// Package enrich defines the structured, AI-derived field model for a job:
// the typed contract for the jobs.enrichment JSONB payload and the controlled
// vocabularies that pin down every enum field's allowed values.
//
// This package is the schema's source of truth. It contains no AI calls — only
// the contract. A later enrichment layer marshals an Enrichment into the JSONB
// column; a later search layer facets on these exact values. Keeping the
// vocabularies here as one definition prevents value fragmentation (e.g.
// "senior" vs "Senior" vs "sr") across those phases.
//
// Field optionality: every field is optional and omitted when the source does
// not state it. Fields whose zero value can be a real value (ints, bool) are
// pointers so an absent field is distinguishable from a present zero; fields
// whose zero value (empty string / empty slice) can never be a valid value use
// omitempty directly.
package enrich

import (
	"fmt"
	"slices"
)

// Enrichment is the typed view of a job's enrichment JSONB payload. JSON keys
// are snake_case to match the existing jobs JSON tags. The blob maps 1:1 to the
// future search document.
type Enrichment struct {
	// Work arrangement.
	WorkMode        string `json:"work_mode,omitempty"`        // enum: WorkModeValues
	EmploymentType  string `json:"employment_type,omitempty"`  // enum: EmploymentTypeValues
	Relocation      string `json:"relocation,omitempty"`       // enum: RelocationValues
	VisaSponsorship *bool  `json:"visa_sponsorship,omitempty"` // pointer: false is meaningful

	// Location / eligibility. Regions is a remote role's geographic reach — a flat,
	// mixed-level vocabulary (global / macro-region / select country). It is
	// meaningful only when WorkMode is "remote". Empty means *unknown*; "global"
	// (open anywhere) is an explicit value, never inferred, so global ≠ unknown.
	Regions      []string `json:"regions,omitempty"`       // enum[]: RegionValues
	Countries    []string `json:"countries,omitempty"`     // enum[]: ISO 3166-1 alpha-2
	Cities       []string `json:"cities,omitempty"`        // free text (not faceted)
	TimezoneNote string   `json:"timezone_note,omitempty"` // free text (not faceted)

	// Compensation.
	SalaryMin      *int   `json:"salary_min,omitempty"`      // in salary_currency units
	SalaryMax      *int   `json:"salary_max,omitempty"`      // in salary_currency units
	SalaryCurrency string `json:"salary_currency,omitempty"` // ISO 4217 (e.g. USD, EUR)
	SalaryPeriod   string `json:"salary_period,omitempty"`   // enum: SalaryPeriodValues

	// Requirements / qualifications.
	Seniority          string   `json:"seniority,omitempty"`            // enum: SeniorityValues
	ExperienceYearsMin *int     `json:"experience_years_min,omitempty"` // non-negative
	EnglishLevel       string   `json:"english_level,omitempty"`        // enum: EnglishLevelValues
	EducationLevel     string   `json:"education_level,omitempty"`      // enum: EducationLevelValues
	Skills             []string `json:"skills,omitempty"`               // normalized lowercase tokens

	// Classification.
	Category        string   `json:"category,omitempty"`         // enum: CategoryValues
	Domains         []string `json:"domains,omitempty"`          // enum[]: DomainValues
	PostingLanguage string   `json:"posting_language,omitempty"` // ISO 639-1 (e.g. en, uk, ru)

	// Company descriptors (job-time observation; seam to the companies entity).
	CompanyType string `json:"company_type,omitempty"` // enum: CompanyTypeValues
	CompanySize string `json:"company_size,omitempty"` // enum: CompanySizeValues
}

// Controlled vocabularies. Each is the ordered, canonical list of allowed
// values for one enum field. They are exported so a later enrichment prompt and
// a later facet config reference the same lists. ISO-standard fields
// (countries, salary_currency, posting_language) and the open skills field have
// no bundled closed vocabulary here and are not enum-validated in this phase.
var (
	WorkModeValues = []string{"remote", "hybrid", "onsite"}
	// RegionValues is the remote-reach vocabulary: global, macro-regions, and a few
	// countries treated as reach areas (extend as the curated facet grows).
	RegionValues = []string{
		"global", "eu", "emea", "eea", "uk", "americas",
		"north_america", "latam", "apac", "mena", "africa", "us", "ru",
	}
	EmploymentTypeValues = []string{"full_time", "part_time", "contract", "internship"}
	RelocationValues     = []string{"not_supported", "supported", "required"}
	SalaryPeriodValues   = []string{"year", "month", "day", "hour"}
	SeniorityValues      = []string{"intern", "junior", "middle", "senior", "lead", "principal", "c_level"}
	EnglishLevelValues   = []string{"none", "a1", "a2", "b1", "b2", "c1", "c2", "native"}
	EducationLevelValues = []string{"none", "bachelor", "master", "phd"}
	CategoryValues       = []string{
		"backend", "frontend", "fullstack", "mobile", "devops", "sre",
		"data_engineering", "data_science", "data_analytics", "ml_ai",
		"qa", "security", "hardware", "embedded", "blockchain",
		"design", "product", "project_management", "management",
		"marketing", "sales", "support", "other",
	}
	DomainValues = []string{
		"fintech", "gambling", "ecommerce", "crypto", "healthcare",
		"saas", "gamedev", "edtech", "adtech", "govtech",
		"media", "travel", "logistics", "other",
	}
	CompanyTypeValues = []string{"product", "startup", "outsource", "outstaff", "agency", "inhouse", "government"}
	CompanySizeValues = []string{"1-10", "11-50", "51-200", "201-500", "501-1000", "1000+"}
)

// Validate checks every enum field against its controlled vocabulary and
// returns an error identifying the first offending field. Empty (absent) fields
// pass — every field is optional. Non-enum fields (ISO codes, free text,
// numbers, skills) are unconstrained here. Multi-value enum fields are checked
// element by element.
func (e Enrichment) Validate() error {
	// Single-value enum fields, in declaration order.
	scalars := []struct {
		field string
		value string
		vocab []string
	}{
		{"work_mode", e.WorkMode, WorkModeValues},
		{"employment_type", e.EmploymentType, EmploymentTypeValues},
		{"relocation", e.Relocation, RelocationValues},
		{"salary_period", e.SalaryPeriod, SalaryPeriodValues},
		{"seniority", e.Seniority, SeniorityValues},
		{"english_level", e.EnglishLevel, EnglishLevelValues},
		{"education_level", e.EducationLevel, EducationLevelValues},
		{"category", e.Category, CategoryValues},
		{"company_type", e.CompanyType, CompanyTypeValues},
		{"company_size", e.CompanySize, CompanySizeValues},
	}
	for _, s := range scalars {
		if s.value != "" && !slices.Contains(s.vocab, s.value) {
			return fmt.Errorf("enrich: invalid %s %q", s.field, s.value)
		}
	}

	// Multi-value enum fields, in declaration order.
	multi := []struct {
		field  string
		values []string
		vocab  []string
	}{
		{"regions", e.Regions, RegionValues},
		{"domains", e.Domains, DomainValues},
	}
	for _, m := range multi {
		for _, v := range m.values {
			if !slices.Contains(m.vocab, v) {
				return fmt.Errorf("enrich: invalid %s %q", m.field, v)
			}
		}
	}

	return nil
}

// Sanitize drops enum values the model emitted outside their controlled
// vocabulary: a scalar field is blanked, a multi-value field keeps only known
// members. This salvages the rest of an otherwise-good payload instead of
// dead-lettering the whole job over one stray value (the model occasionally
// invents a category/region no matter how the vocabulary grows). The invariant
// "never persist an out-of-vocabulary value" still holds — the value is dropped,
// not stored — so Validate passes afterwards.
func (e *Enrichment) Sanitize() {
	scalars := []struct {
		value *string
		vocab []string
	}{
		{&e.WorkMode, WorkModeValues},
		{&e.EmploymentType, EmploymentTypeValues},
		{&e.Relocation, RelocationValues},
		{&e.SalaryPeriod, SalaryPeriodValues},
		{&e.Seniority, SeniorityValues},
		{&e.EnglishLevel, EnglishLevelValues},
		{&e.EducationLevel, EducationLevelValues},
		{&e.Category, CategoryValues},
		{&e.CompanyType, CompanyTypeValues},
		{&e.CompanySize, CompanySizeValues},
	}
	for _, s := range scalars {
		if *s.value != "" && !slices.Contains(s.vocab, *s.value) {
			*s.value = ""
		}
	}

	e.Regions = keepKnown(e.Regions, RegionValues)
	e.Domains = keepKnown(e.Domains, DomainValues)
}

// keepKnown returns values restricted to those present in vocab, preserving order;
// it returns nil when nothing survives so the field omits cleanly.
func keepKnown(values, vocab []string) []string {
	var kept []string
	for _, v := range values {
		if slices.Contains(vocab, v) {
			kept = append(kept, v)
		}
	}
	return kept
}
